client/src/main/java/io/pravega/client/stream/impl/StreamSegments.java

```diff
@@ -67,19 +71,28 @@ public Segment getSegmentForKey(double key) {
         return segments.values();
     }
     
-    public StreamSegments withReplacementRange(StreamSegmentsWithPredecessors replacementRanges) {
+    public StreamSegments withReplacementRange(Segment replacedSegment, StreamSegmentsWithPredecessors replacementRanges) {
+        Preconditions.checkState(segments.containsValue(replacedSegment), "Segment to be replaced should be present in the segment list");
         verifyReplacementRange(replacementRanges);
         NavigableMap<Double, Segment> result = new TreeMap<>();
         Map<Long, List<SegmentWithRange>> replacedRanges = replacementRanges.getReplacementRanges();
-        for (Entry<Double, Segment> exitingSegment : segments.entrySet()) {
-            List<SegmentWithRange> replacements = replacedRanges.get(exitingSegment.getValue().getSegmentId());
-            if (replacements == null || replacements.isEmpty()) {
-                result.put(exitingSegment.getKey(), exitingSegment.getValue());
+        Optional<List<SegmentWithRange>> replacements = Optional.ofNullable(replacedRanges.get(replacedSegment.getSegmentId()));
+        Segment lastSegmentValue = null;
+        for (Entry<Double, Segment> exitingEntry : segments.descendingMap().entrySet()) { // iterate from the highest key.
+            final Segment exitingSegment = exitingEntry.getValue();
+            if (exitingSegment.equals(lastSegmentValue)) {
```

Comment:
Why would we see the same segment twice?

Reply:
If the previous iteration of the loop already added this segment because that iteration of the loop was dealing with the second segment to seal following a merge.